ee/tabby-webserver/src/service/event_logger.rs

```diff
@@ -0,0 +1,189 @@
+        db.create_user("testuser".into(), "pass".into(), true)
+            .await
+            .unwrap();
+
+        logger.log(Event::Completion {
+            completion_id: "test_id".into(),
+            language: "rust".into(),
+            prompt: "testprompt".into(),
+            segments: None,
+            choices: vec![],
+            user: Some("testuser".into()),
+        });
+
+        sleep_50().await;
+        assert!(db.fetch_one_user_completion().await.unwrap().is_some());
+
+        logger.log(Event::View {
+            completion_id: "test_id".into(),
+            choice_index: 0,
+            view_id: None,
+        });
+
+        sleep_50().await;
+        assert_eq!(
+            db.fetch_one_user_completion().await.unwrap().unwrap().views,
+            1
+        );
+
+        logger.log(Event::Dismiss {
+            completion_id: "test_id".into(),
+            choice_index: 0,
+            view_id: None,
+            elapsed: None,
+        });
+
+        sleep_50().await;
+        assert_eq!(
+            db.fetch_one_user_completion()
+                .await
+                .unwrap()
+                .unwrap()
+                .dismisses,
+            1
+        );
+
+        logger.log(Event::Select {
+            completion_id: "test_id".into(),
+            choice_index: 0,
+            view_id: None,
+            kind: None,
+            elapsed: None,
+        });
+
+        sleep_50().await;
+        assert_eq!(
+            db.fetch_one_user_completion()
+                .await
+                .unwrap()
+                .unwrap()
+                .selects,
+            1
+        );
+    }
+
+    #[tokio::test]
+    async fn test_event_logging_no_user() {
+        let db = DbConn::new_in_memory().await.unwrap();
+        let logger = new_event_logger(db.clone());
+
+        logger.log(Event::Completion {
+            completion_id: "test_id".into(),
+            language: "rust".into(),
+            prompt: "testprompt".into(),
+            segments: None,
+            choices: vec![],
+            user: Some("testuser".into()),
+        });
+
+        sleep_50().await;
+        assert!(db.fetch_one_user_completion().await.unwrap().is_none());
+
+        logger.log(Event::Completion {
+            completion_id: "test_id".into(),
+            language: "rust".into(),
+            prompt: "testprompt".into(),
+            segments: None,
+            choices: vec![],
+            user: None,
+        });
+
+        sleep_50().await;
+        assert!(db.fetch_one_user_completion().await.unwrap().is_none());
+
+        logger.log(Event::ChatCompletion {
+            completion_id: "test_id".into(),
+            input: vec![],
+            output: Message {
+                role: "user".into(),
+                content: "test".into(),
+            },
+        });
+
+        sleep_50().await;
+        assert!(db.fetch_one_user_completion().await.unwrap().is_none());
```

Comment:
Consider extract this into a new unit test like `test_chat_completion_event_will_be_skipped`